src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java

```diff
@@ -0,0 +1,101 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package picard.sam.markduplicates;
+
+import htsjdk.samtools.DuplicateSet;
+import htsjdk.samtools.DuplicateSetIterator;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.util.CloseableIterator;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.counting;
+
+public class UmiAwareDuplicateSetIterator extends UmiGraph implements CloseableIterator<DuplicateSet> {
+    private final DuplicateSetIterator wrappedIterator;
+    private Iterator<DuplicateSet> nextSetsIterator;
+    private final int editDistanceToJoin;
+    private final boolean addInferredUmi;
+    private final String umiTag;
+    private final String inferredUmiTag;
+
+    public UmiAwareDuplicateSetIterator(final DuplicateSetIterator wrappedIterator, final int editDistanceToJoin, final boolean addInferredUmi, final String umiTag, final String inferredUmiTag) {
+        this.wrappedIterator = wrappedIterator;
+        this.editDistanceToJoin = editDistanceToJoin;
+        this.addInferredUmi = addInferredUmi;
+        this.umiTag = umiTag;
+        this.inferredUmiTag = inferredUmiTag;
+        nextSetsIterator = Collections.emptyIterator();
+    }
+
+    @Override
+    public void close() {
+        wrappedIterator.close();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return nextSetsIterator.hasNext() || wrappedIterator.hasNext();
+    }
+
+    @Override
+    public DuplicateSet next() {
+        if(!nextSetsIterator.hasNext())
+            process(wrappedIterator.next());
+
+        return nextSetsIterator.next();
+    }
+
+    // Takes a duplicate set and breaks it up into possible smaller sets according to the UMI,
+    // and updates nextSetsIterator to be an iterator on that set of DuplicateSets.
+    private void process(final DuplicateSet set) {
+
+        final List<SAMRecord> records = set.getRecords();
+
+        // If any records are missing the UMI_TAG proceed as if there were no UMIs
+        // and return nextSetsIterator without breaking it up into smaller sets.
+        for(SAMRecord rec : records) {
+            if(rec.getStringAttribute(umiTag) == null) {
+                nextSetsIterator = Collections.singleton(set).iterator();
+                return;
+            }
+        }
+
+        // Count all the uniquely observed UMI sequences
+        Map<String, Long> umiCounts = records.stream()
+                        .collect(Collectors.groupingBy(p -> p.getStringAttribute(umiTag),
+                                counting()));
```

Comment:
@nh13 Would it be better if I made it in a loop rather than used a lambda?
This is just counting the number of times I see each particular UMI.

I've recently changed it to be,
        // Count the number of times each UMI occurs.  If the UMI is null, assume instead that it is ""
        umiCounts = records.stream().map(p -> {final String v = p.getStringAttribute(umiTag); return v==null?"":v;})
            .collect(Collectors.groupingBy(Function.identity(), counting()));

Thoughts?